server/scheduling/scheduler.ts

```diff
@@ -0,0 +1,39 @@
+import { initManagers } from '../db/initDB';
+import { IPost } from '../db/models';
+import mongoose = require('mongoose');
+
+// eslint-disable-next-line no-unused-vars
+const { jobManager, postManager } = initManagers();
+const lt = require('long-timeout');
+
+// run all the tasks which are in the database
+(async() => {
+	(await jobManager.getAllJobs()).forEach((elem) => {
+		if (dateDiff(new Date(), elem.data.date) < 0) {
+			jobManager.deleteJob(elem._id);
+		} else {
+			// eslint-disable-next-line no-new-func
+			const fn = new Function('return ' + elem.value)();
+			scheduleJob(elem.data, fn, elem._id);
+		}
+	});
+})();
+
+function dateDiff(dateOne: Date, dateTwo: Date) : number {
+	return dateTwo.getTime() - dateOne.getTime();
+}
+
+export async function scheduleJob(postData: Partial<IPost>, fn:Function, id?: mongoose.Types.ObjectId): Promise<void> {
+	const time = dateDiff(new Date(), postData.date);
+	if (time >= 0) {
+		if (!id) {
+			id = await jobManager.addJob(postData, fn);
+		}
+
+		lt.setTimeout(() => {
+			// as fn is called with eval we have to add the data to its local scope
+			fn(postManager, postData);
```

Comment:
What will happen if `addJob` throws an error? You don't handle errors and you don't use await as well, so the job will be already deleted by the moment you get an error.

Reply:
Right now if the addJob throws error, the job won't be saved, won't be run (so post will be lost), and (since you don't throw error from catch) nobody will know even that the post is lost (so user cannot submit it again). Check this out, please.

Reply:
And `fn` is `addPost` to DB. So it also doesn't have try catch there or await here. It means that if `fn` throws error you won't wait to check this, and will `deleteJob` earlier.